qpmodel/optimizer.cs

```diff
@@ -292,54 +367,144 @@ public void CountMembers(out int clogic, out int cphysic)
+        // enforce physical property among members
+        // a considered physical node should either supply or can propogate the property
+        public void PropagateProperty(PhysicProperty property)
+        {
+            // after exploration done
+            Debug.Assert(explored_);
+            Debug.Assert(exprList_.Count >= 2);
+
+            if (property is null)
+            {
+                for (int i = 0; i < exprList_.Count; i++)
+                {
+                    if (exprList_[i].physic_ != null && exprList_[i].physic_.RequiredProperty() != null)
+                    {
+                        var child = exprList_[i].physic_.child_();
+                        var group = (child as PhysicMemoRef).Group();
+                        group.PropagateProperty(exprList_[i].physic_.RequiredProperty());
+                    }
+                }
+                return;
+            }
+
+            int hashcode = property.GetHashCode();
+            if (propMember_.ContainsKey(hashcode))
+                return;
+
+            propMember_.Add(hashcode, new List<int>());
+            // loop through all members to find ones with property
+            for (int i = 0; i < exprList_.Count; i++)
+            {
+                if (exprList_[i].physic_ != null)
+                {
+                    if (property.IsPropertySupplied(exprList_[i].physic_) == PropertySupply.supply)
+                    {
+                        propMember_[hashcode].Add(i);
+                        var child = exprList_[i].physic_.child_();
+                        var group = (child as PhysicMemoRef).Group();
+                        group.PropagateProperty(exprList_[i].physic_.RequiredProperty());
+                    }
+                    else if (property.IsPropertySupplied(exprList_[i].physic_) == PropertySupply.propagate)
+                    {
+                        propMember_[hashcode].Add(i);
+                        // TODO: enable multiple child propagation
+                        var child = exprList_[i].physic_.child_();
+                        Debug.Assert(child is PhysicMemoRef);
+                        var group = (child as PhysicMemoRef).Group();
+                        group.PropagateProperty(property);
+                    }
+                }
+            }
+        }
+
+        // when the property is not supplied, the property is enforced
+        // corrresponding physic node is added on top
+        public CGroupMember EnforceProperty(CGroupMember nullmin, PhysicProperty property)
+        {
+            var physic = property.Enforcement(nullmin.physic_);
+            return new CGroupMember(physic, this);
+        }
+
+        public CGroupMember locateMinCostMember(PhysicProperty property)
+        {
+            var nullmin = locateMinCostMember();
+            if (property is null)
+                return nullmin;
+            else
+            {
+                if (propMinMember_.ContainsKey(property.GetHashCode()))
```

Comment:
hashCode is a signature of an object, two objects may have the same signature. So you can't use hashCode to uniquely identify an object.